ast/compile_test.go

```diff
@@ -4164,7 +4193,114 @@ func TestCompilerMockFunction(t *testing.T) {
 	p {true with data.test.is_allowed as "blah" }
 	`)
 	compileStages(c, c.rewriteWithModifiers)
-	assertCompilerErrorStrings(t, c, []string{"rego_compile_error: with keyword cannot replace functions"})
+	assertCompilerErrorStrings(t, c, []string{"rego_compile_error: with keyword cannot replace rego functions"})
+}
+
+func TestCompilerMockBuiltinFunction(t *testing.T) {
+	tests := []struct {
+		note          string
+		module, extra string
+		err           string
+	}{
+		{
+			note: "simple valid",
+			module: `package test
+				now() = 123
+				p { true with time.now_ns as now }
+			`,
+		},
+		{
+			note: "simple valid, simple name",
+			module: `package test
+				mock_concat(_, _) = "foo/bar"
+				p { concat("/", input) with concat as mock_concat }
+			`,
+		},
+		{
+			note: "invalid ref: nonexistant",
+			module: `package test
+				p { true with time.now_ns as now }
+			`,
+			err: "rego_compile_error: with keyword replacing built-in function: value must be a reference to a function",
+		},
+		{
+			note: "valid ref: not a function, but arity = 0", // TODO: arity-0 is special: complete rule... are we OK with this?
+			module: `package test
+				now = 1
+				p { true with time.now_ns as now }
+			`,
+		},
+		{
+			note: "invalid ref: not a function, arity > 0",
+			module: `package test
+				http_send = { "body": "nope" }
+				p { true with http.send as http_send }
+			`,
+			err: "rego_compile_error: with keyword replacing built-in function: referenced value must be a function",
+		},
+		{
+			note: "invalid ref: arity mismatch",
+			module: `package test
+				http_send(_, _) = { "body": "nope" }
+				p { true with http.send as http_send }
+			`,
+			err: "rego_compile_error: with keyword replacing built-in function: referenced value function must have same arity (have 2, want 1)",
+		},
+		{
+			note: "invalid ref: value another built-in",
+			module: `package test
+				p { true with http.send as net.lookup_ip_addr }
+			`,
+			err: "rego_compile_error: with keyword replacing built-in function: value must be a reference to a function",
```

Comment:
Could we make a more specific error message for this case? A built-in function is still a function for the outsider.